File: packages/go/sqllsp/connection.go
package sqllsp

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"strings"
	"sync"
)

// Connection 绑定一次 LSP 会话与 DB session。
type Connection struct {
	ID        string
	SessionID string
	ClientID  string
	// SuggestDatabase 来自编辑器当前库上下文（可空，回退会话默认库）；文档级优先。
	SuggestDatabase string
	// SuggestSchema 金仓/PG 系默认 schema（可空）；文档级优先。
	SuggestSchema string
	// Parser 可选：按连接隔离的方言解析器（如达梦兼容模式）。
	Parser DialectParser
	Docs   *docStore

	diagMu     sync.Mutex
	diagCancel map[string]context.CancelFunc
	diagGen    map[string]uint64
}

// Manager 管理 connectionId → Connection。
type Manager struct {
	mu   sync.Mutex
	byID map[string]*Connection
}

// NewManager 创建连接管理器。
func NewManager() *Manager {
	return &Manager{byID: make(map[string]*Connection)}
}

func newConnectionID() string {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		return hex.EncodeToString([]byte("fallback-conn-id"))
	}
	return hex.EncodeToString(b[:])
}

// Open 创建连接。
func (m *Manager) Open(sessionID, clientID, suggestDatabase string) *Connection {
	m.mu.Lock()
	defer m.mu.Unlock()
	c := &Connection{
		ID:              newConnectionID(),
		SessionID:       sessionID,
		ClientID:        clientID,
		SuggestDatabase: strings.TrimSpace(suggestDatabase),
		Docs:            newDocStore(),
		diagCancel:      make(map[string]context.CancelFunc),
		diagGen:         make(map[string]uint64),
	}
	m.byID[c.ID] = c
	return c
}

// UpdateSuggestDatabase 更新连接级补全默认库（无 uri 时的回退）。
func (m *Manager) UpdateSuggestDatabase(id, database, schema string) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	c, ok := m.byID[id]
	if !ok {
		return false
	}
	c.SuggestDatabase = strings.TrimSpace(database)
	if sch := strings.TrimSpace(schema); sch != "" {
		c.SuggestSchema = sch
	}
	return true
}

// Get 查找连接。
func (m *Manager) Get(id string) (*Connection, bool) {
	m.mu.Lock()
	defer m.mu.Unlock()
	c, ok := m.byID[id]
	return c, ok
}

// Close 关闭并移除连接。
func (m *Manager) Close(id string) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	c, ok := m.byID[id]
	if !ok {
		return false
	}
	c.cancelAllDiags()
	c.Docs.clear()
	delete(m.byID, id)
	return true
}

// CloseBySession 关闭绑定该 session 的全部连接。
func (m *Manager) CloseBySession(sessionID string) int {
	m.mu.Lock()
	defer m.mu.Unlock()
	n := 0
	for id, c := range m.byID {
		if c.SessionID == sessionID {
			c.cancelAllDiags()
			c.Docs.clear()
			delete(m.byID, id)
			n++
		}
	}
	return n
}

func (c *Connection) cancelAllDiags() {
	if c == nil {
		return
	}
	c.diagMu.Lock()
	defer c.diagMu.Unlock()
	for uri, cancel := range c.diagCancel {
		if cancel != nil {
			cancel()
		}
		delete(c.diagCancel, uri)
	}
}

File: packages/cpp/sqllsp/src/connection.cpp
#include "niuma/sqllsp/connection.hpp"

#include <chrono>

namespace niuma::sqllsp {
namespace {

std::string NextConnId(std::uint64_t seq) {
  const auto now = std::chrono::steady_clock::now().time_since_epoch().count();
  return "lsp-" + std::to_string(now) + "-" + std::to_string(seq);
}

}  // namespace

void DocumentStore::Put(const std::string& uri, int version, std::string text) {
  std::lock_guard lock(mu_);
  auto& doc = docs_[uri];
  doc.version = version;
  doc.text = std::move(text);
}

bool DocumentStore::Get(const std::string& uri, Document& out) const {
  std::lock_guard lock(mu_);
  const auto it = docs_.find(uri);
  if (it == docs_.end()) {
    return false;
  }
  out = it->second;
  return true;
}

void DocumentStore::Erase(const std::string& uri) {
  std::lock_guard lock(mu_);
  docs_.erase(uri);
}

void DocumentStore::SetSuggestDatabase(const std::string& uri, const std::string& database,
                                       const std::string& schema) {
  std::lock_guard lock(mu_);
  auto& doc = docs_[uri];
  doc.suggest_database = database;
  if (!schema.empty()) {
    doc.suggest_schema = schema;
  }
}

Connection* Manager::Open(const std::string& session_id, const std::string& client_id,
                          const std::string& suggest_database) {
  std::lock_guard lock(mu_);
  const std::string id = NextConnId(++seq_);
  auto conn = std::make_unique<Connection>();
  conn->id = id;
  conn->session_id = session_id;
  conn->client_id = client_id;
  conn->suggest_database = suggest_database;
  conn->suggest_schema = suggest_database;
  auto* raw = conn.get();
  conns_.emplace(id, std::move(conn));
  return raw;
}

Connection* Manager::Get(const std::string& id) {
  std::lock_guard lock(mu_);
  auto it = conns_.find(id);
  if (it == conns_.end()) {
    return nullptr;
  }
  return it->second.get();
}

bool Manager::Close(const std::string& id) {
  std::lock_guard lock(mu_);
  return conns_.erase(id) > 0;
}

bool Manager::UpdateSuggestDatabase(const std::string& id, const std::string& database,
                                    const std::string& schema) {
  std::lock_guard lock(mu_);
  auto it = conns_.find(id);
  if (it == conns_.end()) {
    return false;
  }
  it->second->suggest_database = database;
  if (!schema.empty()) {
    it->second->suggest_schema = schema;
  } else if (!database.empty()) {
    it->second->suggest_schema = database;
  }
  return true;
}

int Manager::CloseBySession(const std::string& session_id) {
  std::lock_guard lock(mu_);
  int n = 0;
  for (auto it = conns_.begin(); it != conns_.end();) {
    if (it->second->session_id == session_id) {
      it = conns_.erase(it);
      ++n;
    } else {
      ++it;
    }
  }
  return n;
}

}  // namespace niuma::sqllsp
